src/Microsoft.AspNetCore.JsonPatch/Adapters/AdapterFactory.cs

```diff
@@ -0,0 +1,39 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.AspNetCore.JsonPatch.Internal;
+using Newtonsoft.Json.Serialization;
+
+namespace Microsoft.AspNetCore.JsonPatch.Adapters
+{
+    /// <summary>
+    /// The default AdapterFactory to be used for resolving <see cref="IAdapter"/>.
+    /// </summary>
+    public class AdapterFactory : IAdapterFactory
```

Comment:
Does it make sense for the `AdapterFactory` to wrap the `IContractResolver`? IE, you can provide an adapter factory or a contract resolver:
- if you customized the contract resolver but not the adapter factory then we create the default adapter factory
- if you customized the adapter factory then you're expect to figure out how you want to handle contracts

Reply:
@rynowak, I see how this would simplify the code by having a single element to pass around, although it seems to me like we would be asking one class to handle two different purposes.  Did you have another goal in mind for consolidating these?

Late in the PR, @HappyNomad suggested we moved the AdpaterFactory to the ApplyTo, rather than being part of the Constructor.  We had different philosophical views on why, but a decision there might impact a decision here.  In addition to understanding the goal behind your suggestion, I wanted to also get your thoughts on that aspect.  Ultimately if we can get this rolled into production it would be great, so I don't mind a little additional refactoring if it will help us move this forward, I just would rather not do it too many more times ;)

Reply:
I was asking to try and reason about whether or not they are the came concern. The current adapter factory implementation is only thing in the system that interacts with contracts current (I think).

If we were implementing this system fresh (without JSON.NET) we would put the details about how to traverse different types on our equivalent of `JsonContract`.

I don't have any major concerns or strong feedback about what you and @HappyNomad have been discussing, I think both proposals are pretty reasonable. We don't have the luxury of designing something new here because it has to fit into the existing APIs.

Reply:
The concrete implementations of IAdapter are what actually use the ContractResolver.  The AdapterFactory only uses it to pass through.  You are right that we probably could pull it into the AdapterFactory, but it would create more complexity to keep the interface backwards compatible.  It looks like (quick review) that the ContractResolver can be added in various ways (constructor, converter, ApplyTo, etc) so we would need to handle the old mechanism and the new mechanism in all these locations.  I agree that building it fresh it probably makes sense to have the AdpaterFactory provided it if needed.  I am willing to refactor if that will help us get this merged, otherwise I am inclined to make fewer changes than are needed to meet the primary goal of this PR.

Reply:
Ok great, I'm convinced.

Reply:
> Late in the PR, @HappyNomad suggested we moved the AdpaterFactory to the ApplyTo, rather than being part of the Constructor.

I suggested that from the beginning, although I did thoroughly consider your idea to instead accept it in `JsonPatchDocument`'s constructor.  I concluded to stand by my original idea to only accept it in `ApplyTo`.  I summarized my reasoning at https://github.com/aspnet/Home/issues/2816#issuecomment-374306858 and we discussed it starting at https://github.com/aspnet/JsonPatch/pull/132#issuecomment-374069468 .

`IAdapterFactory` is a means of customizing `ObjectAdapter` which is a means of customizing `JsonPatchDocument`.  The existing design is to pass `ObjectAdapter` to `ApplyTo`.  You want to allow the customization of `ObjectAdapter` (`IAdapterFactory`) to be passed to `JsonPatchDocument`'s constructor while `ObjectAdapter` is still passed to `ApplyTo`?  IMHO that's making things more messy and for no good reason.

Please delegate the idea to bake custom behavior into a `JsonPatchDocument` instance to a separate issue rather than conflating it here.  In that separate issue, I think you should consider a `JsonPatchDocument` constructor that accepts the more general `IObjectAdapter` instead of the more specific `IAdapterFactory`.

Please keep this pull request focused and don't touch `JsonPatchDocument` or `JsonPatchDocument<T>`.  Both of our scenarios will work fine without touching those classes.  The extraneous changes you suggest are unnecessary for fulfilling both of our needs.  They may also benefit from further consideration in a separate issue.

Reply:
> We don't have the luxury of designing something new here because it has to fit into the existing APIs.

That's exactly my point.  `IObjectAdapter` is already being passed to `ApplyTo`, so let's not pretend we can undo that design decision by passing `IAdapterFactory` to `JsonPatchDocument`'s constructor.  Instead, let's be consistent with the existing API and leave `JsonPatchDocument` alone.